src/Halogen/ECharts.purs

```diff
@@ -31,65 +32,72 @@ import ECharts.Monad as EM
 import ECharts.Types.Phantom as ETP
 
 import Halogen as H
-import Halogen.HTML.CSS.Indexed (style)
-import Halogen.HTML.Indexed as HH
-import Halogen.HTML.Properties.Indexed as HP
+import Halogen.HTML.CSS (style)
+import Halogen.HTML as HH
+import Halogen.HTML.Properties as HP
+
 
 type EChartsState =
-  { element ∷ Maybe HTMLElement
-  , chart ∷ Maybe ET.Chart
+  { chart ∷ Maybe ET.Chart
   , width ∷ Int
   , height ∷ Int
   }
 
-initialEChartsState ∷ Int → Int → EChartsState
-initialEChartsState w h =
-  { element: Nothing
-  , chart: Nothing
-  , width: w
-  , height: h
-  }
-
 data EChartsQuery a
-  = SetElement (Maybe HTMLElement) a
-  | Init a
+  = Init a
   | Dispose a
   | Set (EM.DSL ETP.OptionI) a
   | Reset (EM.DSL ETP.OptionI) a
   | Resize a
   | Clear a
-  | SetHeight Int a
-  | SetWidth Int a
+  | SetDimensions { width ∷ Maybe Int, height ∷ Maybe Int } a
   | GetOptions (Maybe Foreign → a)
   | GetWidth (Int → a)
   | GetHeight (Int → a)
 
+data EChartsMessage = Initialized
+
 type EChartsEffects eff =
   ( echarts ∷ ET.ECHARTS
   , dom ∷ DOM
   , avar ∷ AVAR
   , err ∷ EXCEPTION
+  , ref ∷ REF
   | eff
   )
 
+
 type HTML = H.ComponentHTML EChartsQuery
-type DSL g = H.ComponentDSL EChartsState EChartsQuery g
+
+type DSL g = H.ComponentDSL EChartsState EChartsQuery EChartsMessage g
 
 echarts
-  ∷ ∀ eff g
-  . (Affable (EChartsEffects eff) g)
-  ⇒ H.Component EChartsState EChartsQuery g
-echarts = H.lifecycleComponent
-  { render
+  ∷ ∀ eff g r
+  . (MonadAff (EChartsEffects eff) g)
+  ⇒ { width ∷ Int, height ∷ Int | r}
+  → H.Component HH.HTML EChartsQuery Unit EChartsMessage g
+echarts = echarts' $ const Nothing
+
+echarts'
+  ∷ ∀ eff g i r
+  . (MonadAff (EChartsEffects eff) g)
+  ⇒ (i → Maybe (EChartsQuery Unit))
+  → { width ∷ Int, height ∷ Int | r }
```

Comment:
Maybe we should make the dimensions the input value for the component? Seems like a prime candidate for where that kind of thing would be great. The parameterised input thing here is pretty cool, but I think dimensions is by far the most useful possible thing we could use `i` for (it was actually the motivating case for why I added inputs to this release) - manual querying seems more appropriate for updating with new data, etc.

I guess we could do a mixture of the two, if we used `Tuple Dimensions i -> ...` perhaps?